src/ophyd_async/core/_table.py

```diff
@@ -64,90 +88,44 @@ def __add__(self, right: TableSubclass) -> TableSubclass:
+        return super().__eq__(value)
+
+    def numpy_dtype(self) -> np.dtype:
         dtype = []
-        for field_name, field_value in cls.model_fields.items():
-            if np.ndarray in (
-                get_origin(field_value.annotation),
-                field_value.annotation,
-            ):
-                dtype.append((field_name, getattr(cls, field_name).dtype))
+        for k, v in self:
+            if isinstance(v, np.ndarray):
+                dtype.append((k, v.dtype))
             else:
-                enum_type = get_args(field_value.annotation)[0]
-                assert issubclass(enum_type, Enum)
-                enum_values = [element.value for element in enum_type]
-                max_length_in_enum = max(len(value) for value in enum_values)
-                dtype.append((field_name, np.dtype(f"<U{max_length_in_enum}")))
-
+                # TODO: use np.dtypes.StringDType when we can use in structured arrays
+                # https://github.com/numpy/numpy/issues/25693
+                dtype.append((k, np.dtype("S40")))
         return np.dtype(dtype)
 
-    def numpy_table(self):
-        # It would be nice to be able to use np.transpose for this,
-        # but it defaults to the largest dtype for everything.
-        dtype = self.numpy_dtype()
-        transposed_list = [
-            np.array(tuple(row), dtype=dtype)
-            for row in zip(*self.numpy_columns(), strict=False)
-        ]
-        transposed = np.array(transposed_list, dtype=dtype)
-        return transposed
-
-    def numpy_columns(self) -> list[np.ndarray]:
-        """Columns in the table can be lists of string enums or numpy arrays.
-
-        This method returns the columns, converting the string enums to numpy arrays.
-        """
-
-        columns = []
-        for field_name, field_value in self.model_fields.items():
-            if np.ndarray in (
-                get_origin(field_value.annotation),
-                field_value.annotation,
-            ):
-                columns.append(getattr(self, field_name))
-            else:
-                enum_type = get_args(field_value.annotation)[0]
-                assert issubclass(enum_type, Enum)
-                enum_values = [element.value for element in enum_type]
-                max_length_in_enum = max(len(value) for value in enum_values)
-                dtype = np.dtype(f"<U{max_length_in_enum}")
-
-                columns.append(
-                    np.array(
-                        [enum.value for enum in getattr(self, field_name)], dtype=dtype
-                    )
-                )
-
-        return columns
+    def numpy_table(self, selection: slice | None = None) -> np.ndarray:
+        array = None
+        for k, v in self:
+            if selection:
+                v = v[selection]
+            if array is None:
+                array = np.empty(v.shape, dtype=self.numpy_dtype(self))
+            array[k] = v
+        assert array
+        return array
```

Comment:
I think we can do:

```suggestion
    def numpy_table(self, selection: slice = [::]) -> np.ndarray:
        return np.array(
            [
                (k, v[selection])
                for k, v in self
            ],
            dtype=self.numpy_dtype()
        )
```

Has the perk of vacuously working with empty tables.

Reply:
Ah I'm incorrect, ignore me